### ISSUE TITLE
Reduce footprint of distribution?

### ISSUE THREAD

**Issue by wendellpiez** [djbpitt/docker-basex#4]
Currently, BaseX materials are included in the distribution, because when they weren't, the configuration was failing. However, particularly in the `webapp` subdirectory and possibly also in `repo`, the distribution has copies of things including BaseX demonstrations, which perhaps do not actually need to be present for things to work.

And these copies should really not be provided in the repo; the question is whether we can configure BaseX so that its own (for example) `dba` application might be included along with our demo application (in `iching`), without our having to have a copy but instead relying on the image to provide it.

This is a Docker question - about how and when things must be included, or what can be inherited from images.

**Comment by djbpitt** [comment]
@wendellpiez I think it's also a BaseX issue. If BaseX is like eXist-db in this respect, all content that has been installed into the DB is stored in a proprietary format that does not permit separating applications, and that content includes things like the user files, with passwords, groups, etc., so it isn't just what we think of as installed apps. I think means that if we want the user to be able to fire up the container from a Docker image and interact with it, including changing the DB contents and persisting the changes through a shutdown cycle, there are two options: persist the changes in the container after shutdown or store the changed (or changeable) files on the filesystem and mount them on launch. 

There's a discussion of this at <https://stackoverflow.com/questions/28574433/do-docker-containers-retain-file-changes>. It sounds as if both approaches are possible, but treating containers as read-only is considered Best Practice, and if we want to follow that, BaseX DB files that can change (in other words, much of the DB) would need to be stored on the local filesystem. 

I don't think we can include BaseX core demos alongside our own demos by putting theirs in the Docker image and ours on the local filesystem. If I've understood correctly, all content uploaded into the DB gets blended together into the same binary files. Either that puree all goes inside the image and we persist work-session changes there, as well, or it all goes on the filesystem.

The preceding reflects a pretty thin understanding of how Docker operates, so it may all be entirely wrong. But I don't think it is.

**Comment by wendellpiez** [comment]
I don't think it's entirely wrong but it's not entirely right either. There are really two separate issues -- which resources BaseX needs to run and how to see to it they are in the container (without installing BaseX separately); and how to persist 'state' on the system (which turns out to be, by externalizing it: and this in fact works fine for me). I think we can separate these -- indeed the second one (at least giving what we are trying to do so far) is not so much of a problem as the first. As for the first question, going for workable now and optimal later, is also okay -- although there are also some easy experiments we could try.

**Comment by djbpitt** [comment]
@wendellpiez As I slept on the idea, it occurred to me that although I think it is possible to persist information inside the container without writing it to the file system (that is, if the container is stopped and then restarted, the changes will be there), the container is created from an image, so if the user removes (doesn’t just shut down) the container and then launches a new one from the same image, any changes will be lost. That’s consistent with your observation, above, that the only robust way to persist  user-specific information is on the file system, where it’s the responsibility of the user, and not of the container environment. 

For that reason (barring any misunderstanding on my part about how Docker operates), I agree with the suggestion in your second paragraph that user data that is to be loaded into BaseX would be loaded from the filesystem on launch of the container, although I’m not confident that I’ve thought of all of the likely use cases. In particular, I don’t know whether the best approach would involve storing the binary DB files on the filesystem or exporting them in serialized form and reloading them into the DB on launch. And probably other options I’m not considering yet. But following also your suggestion about “workable now and optimal later”, I think we can defer further consideration of options, and just settle on one that works for the moment.

**Comment by wendellpiez** [comment]
@djbpitt Exactly. The way I am thinking of a container is a game environment all set up ready to play. The image is like the game in its box. Run the container, and the game starts the same way every time. If you shut it down, you can't come back to it, you have to restart. Of course once we start persisting data or processes outside the container (note it is a feature of the architecture that not only source data but also the XQuery and XSLT can be exposed this way), things are a bit different, there is that state to consider. But that is home ground for us.

**Comment by djbpitt** [comment]
@wendellpiez Pursuing the Game In The Box analogy, if I’ve correctly understood how containers work (and I may well not have), Docker distinguishes two types of *shutting down* a container:

1. `stop` pauses a container. A stopped container can be restarted and it’s in the same state as it was in when when we stopped it. That means that we can upload content into the DB, stop the container, and the content will be there when we return. This is like walking away from a game in progress without putting the pieces back in the box.
1. `kill` deletes a container, which removes it entirely. It cannot be unkilled, but a new container can be started from the same image. This is like putting a game away and then taking the box out again later to play it from the beginning.

The `--rm` switch to `docker run` in our *lunch.sh* automatically deletes the container when we stop it. But if we want it to remain so that we can restart it, we can remove that switch. In that case, we can exit out of our `docker run` with Ctrl-c, as we usually do, but that just stops the container without deleting it, so that we can restart the same container (walk back to the open game board) with `docker start xmllunchbox`. If we do that, by the way, we won’t be allowed to run our *lunch.sh* script again; it will give us an error that it cannot create a new container called *xmllunchbox* because there is already a stopped container with that name.

Containers are launched from images, so in addition to stopping or killing containers, it’s also possible to delete images. When that happens, you can’t launch containers from them until you rebuild them, that us, until you build a new image from the same Dockerfile. You can’t remove an image if it has any associated containers, including stopped containers. The stopped containers have to be removed first.

Thus endeth today’s lesson from the *Book of Docker*.

**Comment by wendellpiez** [comment]
This corresponds to my understanding as well.

As presently configured, the `webapp` directory contains our demo code, plus also the BaseX demo code. Additionally, there is (or is likely to be) code of ours and also of theirs in `repo`. The question is, if we drop their demonstrations, whether they might be already in the image anyway (so we could use them without maintaining copies). Or perhaps the image is set up without them, and if we want BaseX but with their restXQ demos as well as ours, we have no choice but to include them.

The subdirectories include `dba` and `chat` inside `basex/webapp`. (They are also not terribly large if that makes a difference.)
